From ad241a935d2c08e8a8a4943481b0b6d200838959 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Alex=20R=C3=B8nne=20Petersen?= <alex@alexrp.com>
Date: Fri, 17 Oct 2025 14:59:22 +0200
Subject: [PATCH 009/133] libcxx: backport llvm/llvm-project#158347
X-Developer-Signature: v=1; a=openpgp-sha256; l=1179; i=xtex@astrafall.org;
 h=from:subject; bh=ZuJGoNmNdbwzl0eaYlNTsJzuKnV7IFuLnOJTNSiexk8=;
 b=owGbwMvMwCW2U4Ij7wZL9ETG02pJDJn7cn9++7L5t1SziBsXH+sqJ0Ezf8XjfZvv/6o1uPrGY
 kUDx7cnHRNZGMS4GCzFFFmKDBu8WXXS+UWXlcvCzGFlAhkiLdLAwMDAwMLAl5uYV2qkY6Rnqm2o
 Z2ioY6RjxMDFKQBTncjMyPBsZ6TkfJ8E1zk3TGbvU1O4xqLfqijkyXi52tw669CPIi5Ghg+s8Y0
 y22WZGJ5O1LGxzXBbLTmxZ/qZi0uzS9t8Jxd2MwAA
X-Developer-Key: i=xtex@astrafall.org; a=openpgp;
 fpr=7231804B052C670F15A6771DB918086ED8045B91

https://github.com/llvm/llvm-project/pull/158347
(cherry picked from commit 751375f3ca5ba4a3a9edb2e79f69fb6e27a445b8)
---
 lib/libcxx/include/__ranges/join_view.h | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/lib/libcxx/include/__ranges/join_view.h b/lib/libcxx/include/__ranges/join_view.h
index 327b349f476a..364f056d8d2c 100644
--- a/lib/libcxx/include/__ranges/join_view.h
+++ b/lib/libcxx/include/__ranges/join_view.h
@@ -410,8 +410,13 @@ struct __segmented_iterator_traits<_JoinViewIterator> {
 
   static constexpr _LIBCPP_HIDE_FROM_ABI _JoinViewIterator
   __compose(__segment_iterator __seg_iter, __local_iterator __local_iter) {
-    return _JoinViewIterator(
-        std::move(__seg_iter).__get_data(), std::move(__seg_iter).__get_iter(), std::move(__local_iter));
+    auto&& __parent = std::move(__seg_iter).__get_data();
+    auto&& __outer  = std::move(__seg_iter).__get_iter();
+    if (__local_iter == ranges::end(*__outer)) {
+      ++__outer;
+      return _JoinViewIterator(*__parent, __outer);
+    }
+    return _JoinViewIterator(__parent, __outer, std::move(__local_iter));
   }
 };
 
-- 
2.52.0

